### PR TITLE
rolling functions

### PR DESCRIPTION
**Pull request by jangorecki**
draft version for rolling functions requested in #6 

we can define scope for those tests here. Some are already there, others only mentioned in comments

----

PR roadmap:
- [x] test scripts
  - [x] data.table
  - [x] dplyr
  - [x] pandas
  - [x] duckdb-latest
  - [x] spark
- [x] validate results
  - [x] data.table
  - [x] dplyr
  - [x] pandas
  - [x] duckdb-latest
  - [x] spark
- [x] add rollfun to report
- [x] fix https://github.com/duckdblabs/db-benchmark/issues/29
- [x] run rollfun benchmark and confirm rollfun benchplot
- [x] branch installation version logging to logs.csv

optionally: test scripts and validate results:
- juliadf
- polars - @ritchie46 any chance for the script?
- clickhouse - https://clickhouse.com/docs/en/sql-reference/window-functions
- datafussion - @Dandandan https://arrow.apache.org/datafusion/user-guide/sql/window_functions.html
- dask - https://docs.dask.org/en/stable/dataframe-api.html#rolling-operations

rolling functions not available in:
- pydatatable - https://github.com/h2oai/datatable/issues/1500
- juliads
- arrow - https://github.com/r-lib/slider/issues/195

### CODE REVIEW

**Comment by jangorecki** [comment]
As there was no feedback on the scope proposed by me in April, I made another step forward and define it more precisely. Please have a look at the questions proposed. It is not easy, within 10 questions, to cover well all the possible features, so I ended up focusing on:

- funs: mean, median, regression, UDF
- scalability: window small or big (length of the input is not part of the questions, it is parameter of the script, so this scalability is always tested)
- features: vectorized input, weighted fun, uneven time series, by.column=F

```
q1: rolling mean
q2: window small
q3: window big
q4: multi vars+cols
q5: median
q6: weighted
q7: uneven dense
q8: uneven sparse
q9: regression (by.column)
q10: UDF (not decided yet but has to be one that will generally not be optimized to mimic arbitrary UDF)
```

Looking forward for feedback on the scope for that test, or implementations in other software.
Note that we may be forced to move from N {1e6, 1e7, 1e8} to N {1e5, 1e6, 1e7} depending on how well other tools will scale.

**Comment by jangorecki** [comment]
I pushed implementation for dplyr (using slider) as well. So we have at the moment data.table and dplyr.

It is now nicely run-able just by calling `./run.sh`.

I commented out q10 "UDF" for now. It is completely not comparable with other questions, being x1000-10000 slower than the rest. We either have to figure out some very lightweight UDF, or just drop UDF from the scope.

**Comment by jangorecki** [comment]
Instead of q10 UDF I would propose either:

- rolling `min` - still it is quite different implementation than mean/sum
- second holistic aggregate - but here I don't see that much value added - looking at https://duckdb.org/2021/11/12/moving-holistic.html - `mode` is not useful for our `float64` measure variable, `median` is already inside this benchmark, `quantile` is not much different than median, median is actual a special case of quantile.

So I am in favor of `min`.

Posting here before doing the change as I hope there may be some other ideas.

**Comment by jangorecki** [comment]
Other topics that are subject to community review are:
1. data size
  - 1e5, 1e6, 1e7 (could be desired to run slower tools)
  - 1e6, 1e7, 1e8 (currently implemented, feels fine)
  - 1e7, 1e8, 1e9 (if we drop UDF in q10 this should be do-able)
3. window size
```r
w = nrow(x)/1e3       ## used in 8 out of 10 questions
wsmall = nrow(x)/1e4  ## used q2 only
wbig = nrow(x)/1e2    ## used q3 only
```

**Comment by AdrianAntico** [comment]
@jangorecki you have a solid set of measures, the only other type I'd consider would be differencing

**Comment by rdavis120** [comment]
I think it would be helpful if the dplyr test was documented as a representation of the slider package, as opposed to using RcppRoll.  The dplyr benchmark has been confusing to me because you can use dplyr with duckdb or data.table.  For data.table, it is more obvious that you are benchmarking the rolling functions inside the data.table package.

**Comment by jangorecki** [comment]
@AdrianAntico that is an interesting idea, but as I briefly looked at potential implementation, it doesn't seem to stress windowing computation (use of `diff` in R). If you could provide an example of code then it will be more clear.

@rdavis120 maintaining new solutions is relatively high cost, putting slider under dplyr was easy way to avoid that. Rather than adding new solution slider I would prefer to rename dplyr to tidyverse so it will fit well and there will be no need for adding another solution. Anyway I would prefer to keep this issue discussion around rolling task scope rather naming details.

**Comment by AdrianAntico** [comment]
@jangorecki the intent was more time series related (and cross-row related). I have a diff function in my github package "Rodeo" if you want a full example, starting at line 443... https://github.com/AdrianAntico/Rodeo/blob/main/R/FeatureEngineering_CrossRowOperations.R

**Comment by jangorecki** [comment]
@Tmonster could we get CI workflow approval?

**Comment by jangorecki** [comment]
@Tmonster Is there a way that I could be approving CI runs? does it run on duckdblabs private runners? if not then I don't think there should be any concerns.

I added pandas rollfun script, and (hopefully) fixed failure in previous GH Actions job.

**Comment by Tmonster** [comment]
> @Tmonster Is there a way that I could be approving CI runs? does it run on duckdblabs private runners? if not then I don't think there should be any concerns.
> 
> I added pandas rollfun script, and (hopefully) fixed failure in previous GH Actions job.

Yea, I thought the CI would run after every push once I approve it the first time. I think I don't have the permissions to enable that yet. Will talk to mark about it tomorrow 👍

File: duckdb-latest/rollfun-duckdb-latest.R [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+# window size
+w = in_nr/1e3
+wsmall = in_nr/1e4
+wbig = in_nr/1e2
+
+task_init = proc.time()[["elapsed"]]
+cat("rolling...\n")
+
+question = "mean" # q1
+sql = sprintf("CREATE TABLE ans AS SELECT avg(v1) OVER (ORDER BY id1 ROWS BETWEEN %d PRECEDING AND CURRENT ROW) AS v1 FROM x", w-1L)
```
Comment (Dandandan) [inline duckdb-latest/rollfun-duckdb-latest.R:56]: The same queries can be used for DataFusion or spark as well

Comment (jangorecki) [reply duckdb-latest/rollfun-duckdb-latest.R:56]: that's great. unfortunately my volunteering availability is limited at the moment, and adding even one similar solution is quite a bit of work... setting up its environment, modifying script, validating results against another tool, checking it passes _report/report.R checks. If all goes smooth it's couple hours of work, but I don't think it ever went smooth :)

**Comment by bkamins** [comment]
@jangorecki I will propose code for juliadf.

File: duckdb-latest/rollfun-duckdb-latest.R (hunk 1 above)
Comment (Dandandan) [reply duckdb-latest/rollfun-duckdb-latest.R:56]: No worries, just stating that the syntax should be the same for those engines 👍

**Comment by bkamins** [comment]
@jangorecki - which is the reference implementation of all the questions from Q1 to Q10 (I am asking, because I have checked several solutions and they indicate "not implemented yet" and I do not know how to exactly reproduce them in Julia)

**Comment by bkamins** [comment]
But then - back to my original question => which is the reference implementation of the questions I should match the results against in Julia implementation? (in particular - do you want to include constant term in the regression and what should be returned from the operation)

Thank you!

**Comment by jangorecki** [comment]
That haven't been settled yet. I need to go through stackoverflow to find common questions about it. Actually it would be helpful if you could propose one which you believe is the common problem that users are looking to solve with rolling regression.

**Comment by bkamins** [comment]
Incidentally, just today I saw an example from a user. The user wanted to run `y ~ x` kind of regression and keep the result as a vector of collections: Something like:
```
3-element Vector{Vector{Float64}}:
 [0.11728235062958436, 0.9228342578148421]
 [0.2160138268973083, 0.41776928538024183]
 [0.42587771406039454, 0.10348333203334836]
```
(so both intercept and slope are kept in a single entry)

**Comment by jangorecki** [comment]
What I did for duckdb q10 for now is r^2, because this is what we used in groupby q9

**Comment by jangorecki** [comment]
duckdb and spark (and possibly pandas only for q8 and q9) does not have an options for handling properly an incomplete rolling window. Timings for those solutions will not include required postprocessing to match exactly same result: NULL vs value from the unexpected window size.

**Comment by jangorecki** [comment]
Development is possibly finished on this branch. 5 solutions added till now have been validated using https://github.com/jangorecki/db-benchmark/blob/rollfun/_utils/rollfun-ans-validation.txt

@bkamins if you would like to add Julia, you are welcome, please use commands from file linked above to validate answers against one of the solutions.

**Comment by jangorecki** [comment]
@Tmonster PR is ready to merge

To reproduce
```sh
# install R and python

git clone https://github.com/jangorecki/db-benchmark --branch rollfun --single-branch --depth 1
cd db-benchmark
# install solutions interactively
./dplyr/setup-dplyr.sh
./datatable/setup-datatable.sh
./pandas/setup-pandas.sh
./duckdb-latest/setup-duckdb-latest.sh
./spark/setup-spark.sh

# prepare data
Rscript _data/rollfun-datagen.R 1e6 NA 0 1
Rscript _data/rollfun-datagen.R 1e7 NA 0 1
Rscript _data/rollfun-datagen.R 1e8 NA 0 1
mkdir data
mv R1_1e*_NA_0_1.csv data

vim run.conf
# do upgrade false
# force run true
# disable report
# disable publish
# run_task rollfun
# run_solution data.table dplyr pandas duckdb-latest spark

sudo swapoff -a

./run.sh > ./run.out
```

**Comment by jangorecki** [comment]
@Tmonster I recall getting an update on this but cannot find it here and in related issue. Could you please provide the status?

I was about to suggest to push this forward, considering that duckdb advertised this features nicely during duckdb 1.1 release: https://duckdb.org/2024/09/09/announcing-duckdb-110.html

> Window functions see a lot of use in DuckDB, which is why we are continuously improving performance of executing Window functions over large datasets.

**Comment by Tmonster** [comment]
Hi @jangorecki ,

I can’t seem to find our previous discussion either for some reason. As for this PR, It has been our decision at DuckDB Labs to maintain the benchmark suite “as-is” when it comes to the tested functionality. The worry is that if we start expanding the questions the benchmark will develop into a “DuckDB Labs” benchmark, which is something we would like to prevent. You are happy to fork and extend it, however.

**Comment by AdrianAntico** [comment]
@Tmonster is this because DuckDB doesn't do well on these operations? Shouldn't this be a reason to motive your team to improve the performance of these query types vs hiding the performance issue?

image

**Comment by jangorecki** [comment]
@AdrianAntico I think there is no way for duckdb do be competitive in this kind of operations (that depends on the order of data) because it does not have a concept of physical order of data (clustered index). I believe it can be best in class among such tools although it won't really compete with tools that have clustered index. Therefore for time series it may not be ideal solution. Unless clustered index will make its way to duckdb, but that probably required change of data formats.

**Comment by AdrianAntico** [comment]
@jangorecki it's such a common operation (windowing) that it's hard to consider DuckDB as a feasible option for everyday use without it. How was this overlooked from inception?

**Comment by jangorecki** [comment]
It is not really question for me to answer, but I think duckdb aligned to existing sql rdbms'es in that regard which by default does not maintain order. Many of them do have clustered index but OLTP databases don't really need clustered index, it is OLAP that has much more to gain from it. Considering duckdb is OLAP then it feels to be a good FR for duckdb.

**Comment by Mytherin** [comment]
> @Tmonster is this because DuckDB doesn't do well on these operations? Shouldn't this be a reason to motive your team to improve the performance of these query types vs hiding the performance issue?

As explained above - the original reason we decided to host the h2oai db-benchmark is because the original hosted version was no longer being updated, and had very outdated results using old versions of DuckDB and other tools. The goal of this project is to resurrect and maintain the db-benchmark *as-is*, but re-running it with new versions of different tools. 

Extending and changing the benchmark means we are no longer running the h2oai db-benchmark, but rather a *DuckDB Labs customized version of the db-benchmark*. This then comes with a lot of added work and drama - which benchmarks do you add, which do you not - together with potential accusations. What if we add a benchmark for which we perform better than system X? What if we don't add a benchmark where we perform worse than system X?

As such, to avoid this drama, we have opted to only preserve the original benchmark rather than extend it.

We are not trying to hide any performance issues. You are free to run any benchmarks of your own choosing using DuckDB. Our response to benchmarks has always been to take them and use them to improve the system. We have actually recently [greatly improved our windowing lag performance in DuckDB v1.1](https://github.com/duckdb/duckdb/pull/12609#issuecomment-2181323484).

> @AdrianAntico I think there is no way for duckdb do be competitive in this kind of operations (that depends on the order of data) because it does not have a concept of physical order of data (clustered index). I believe it can be best in class among such tools although it won't really compete with tools that have clustered index. Therefore for time series it may not be ideal solution. Unless clustered index will make its way to duckdb, but that probably required change of data formats.

This is not entirely true - and we are working on adding further optimizations that take better advantage of natural order in the data. The reason this hasn't been done in the past is that it is a lot harder to do so when working in a streaming engine that operates on larger-than-memory data, versus when working with in-memory arrays. In addition, the Window operator in SQL is a lot more extensive and complex than what is generally provided in DataFrame libraries.

**Comment by jangorecki** [comment]
Very happy to heard about incoming optimizations. And it is a great news to be able to run window functions on out of memory data! Is there issue for that so we can subscribe and be notified when it will be resolves?

**Comment by jangorecki** [comment]
For those who are wondering about actual timings I presented them earlier this year in my slides.
https://jangorecki.gitlab.io/r-talks/2024-01-26_Edinburgh_Rolling-statistics/Rolling-statistics.pdf

**Pull request closed without merging**